src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,769 @@
+        private static readonly TypeSyntax VarNameIdentifier = SyntaxFactory.IdentifierName("var");
+
+        protected override string Title => CSharpFeaturesResources.Convert_to_foreach;
+
+        protected override bool TryConvert(
+            QueryExpressionSyntax queryExpression,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken,
+            out DocumentUpdateInfo documentUpdateInfo)
+            => Converter.TryConvert(queryExpression, semanticModel, cancellationToken, out documentUpdateInfo);
+
+        private sealed class Converter
+        {
+            private readonly SemanticModel _semanticModel;
+            private readonly CancellationToken _cancellationToken;
+
+            private Converter(SemanticModel semanticModel, CancellationToken cancellationToken)
+            {
+                _semanticModel = semanticModel;
+                _cancellationToken = cancellationToken;
+            }
+
+            public static bool TryConvert(
+                QueryExpressionSyntax queryExpression,
+                SemanticModel semanticModel,
+                CancellationToken cancellationToken,
+                out DocumentUpdateInfo documentUpdateInfo)
+                => new Converter(semanticModel, cancellationToken).TryConvert(queryExpression, out documentUpdateInfo);
+
+            private bool TryConvert(QueryExpressionSyntax source, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                if (source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                    SyntaxKind.SingleLineCommentTrivia,
+                    SyntaxKind.MultiLineCommentTrivia,
+                    SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                    source.ContainsDirectives))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // GetDiagnostics is expensive. Move it to the end if there were no bail outs from the algorithm.
+                // TODO likely adding more semantic checks will perform checks we expect from GetDiagnostics
+                // We may consider removing GetDiagnostics.
+                // https://github.com/dotnet/roslyn/issues/25639
+                if (!TryConvertInternal(source, out documentUpdateInfo) || 
+                    _semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any())
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                return true;
+            }
+
+            private bool TryConvertInternal(QueryExpressionSyntax source, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // (from a in b select a); 
+                SyntaxNode parent = source.WalkUpParentheses().Parent;
+
+                // (from a in b select a)?.ToList();
+                if (parent.Kind() == SyntaxKind.ConditionalAccessExpression)
```

Comment:
>if (parent.Kind() == SyntaxKind.ConditionalAccessExpression) [](start = 16, length = 60)

Instead of walking up conditional access, should we consider to `TryReplaceWithLocalFunction` here so that the user does not lose their null check?